src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -625,27 +626,20 @@ void ShenandoahDirectCardMarkRememberedSet::merge_write_table(HeapWord* start, s
 }
 
 // Destructively copy the write table to the read table, and clean the write table.
-void ShenandoahDirectCardMarkRememberedSet::reset_remset(HeapWord* start, size_t word_count) {
-  size_t start_index = card_index_for_addr(start);
-#ifdef ASSERT
-  // avoid querying card_index_for_addr() for an address past end of heap
-  size_t end_index = card_index_for_addr(start + word_count - 1) + 1;
-#endif
-  assert(start_index % ((size_t)1 << LogCardValsPerIntPtr) == 0, "Expected a multiple of CardValsPerIntPtr");
-  assert(end_index % ((size_t)1 << LogCardValsPerIntPtr) == 0, "Expected a multiple of CardValsPerIntPtr");
-
-  // We'll access in groups of intptr_t worth of card entries
-  intptr_t* const read_table  = (intptr_t*) &(_card_table->read_byte_map())[start_index];
-  intptr_t* const write_table = (intptr_t*) &(_card_table->write_byte_map())[start_index];
-
-  // Avoid division, use shift instead
-  assert(word_count % ((size_t)1 << (LogCardSizeInWords + LogCardValsPerIntPtr)) == 0, "Expected a multiple of CardSizeInWords*CardValsPerIntPtr");
-  size_t const num = word_count >> (LogCardSizeInWords + LogCardValsPerIntPtr);
+void ShenandoahDirectCardMarkRememberedSet::reset_remset() {
+  // iterate on threads and adjust thread local data
+  struct ResetRemSet : public ThreadClosure {
+    CardTable::CardValue* new_ptr;
+    ResetRemSet(CardTable::CardValue* np) {
+      this->new_ptr = np;
+    }
+    virtual void do_thread(Thread* t) {
+      ShenandoahThreadLocalData::set_map_base(t, new_ptr);
+    }
+  } reset_remset (_card_table->swap_bases());
 
-  for (size_t i = 0; i < num; i++) {
-    read_table[i]  = write_table[i];
-    write_table[i] = CardTable::clean_card_row_val();
-  }
+  log_info(gc, remset)("Resetting remset to: " PTR_FORMAT, p2i(reset_remset.new_ptr));
+  Threads::threads_do(&reset_remset);
```

Comment:
Which threads should update?

Reply:
This should `do` all the threads. In some cases, gc threads (and even the vm thread) may perform evacuations and/or promotions that may require updating the remembered set.